components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/device/grant/DeviceFlowGrant.java

```diff
@@ -0,0 +1,154 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.oauth2.device.grant;
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.identity.oauth2.IdentityOAuth2Exception;
+import org.wso2.carbon.identity.oauth2.device.constants.Constants;
+import org.wso2.carbon.identity.oauth2.device.dao.DeviceFlowPersistenceFactory;
+import org.wso2.carbon.identity.oauth2.device.errorcodes.DeviceErrorCodes;
+import org.wso2.carbon.identity.oauth2.device.model.DeviceFlowDO;
+import org.wso2.carbon.identity.oauth2.dto.OAuth2AccessTokenReqDTO;
+import org.wso2.carbon.identity.oauth2.dto.OAuth2AccessTokenRespDTO;
+import org.wso2.carbon.identity.oauth2.model.RequestParameter;
+import org.wso2.carbon.identity.oauth2.token.OAuthTokenReqMessageContext;
+import org.wso2.carbon.identity.oauth2.token.handlers.grant.AbstractAuthorizationGrantHandler;
+import org.wso2.carbon.identity.oauth2.util.OAuth2Util;
+
+import java.sql.Timestamp;
+import java.util.Date;
+
+/**
+ * Device flow grant type for Identity Server.
+ */
+public class DeviceFlowGrant extends AbstractAuthorizationGrantHandler {
+
+    private static Log log = LogFactory.getLog(DeviceFlowGrant.class);
+
+    @Override
+    public boolean validateGrant(OAuthTokenReqMessageContext oAuthTokenReqMessageContext) throws
+            IdentityOAuth2Exception {
+
+        super.validateGrant(oAuthTokenReqMessageContext);
+        OAuth2AccessTokenReqDTO tokenReq = oAuthTokenReqMessageContext.getOauth2AccessTokenReqDTO();
+
+        boolean authStatus = false;
+        RequestParameter[] parameters = oAuthTokenReqMessageContext.getOauth2AccessTokenReqDTO()
+                .getRequestParameters();
+        String DeviceCode = null;
+        String deviceStatus = null;
+
+        for (RequestParameter parameter : parameters) {
+            if (Constants.DEVICE_CODE.equals(parameter.getKey())) {
+                if (parameter.getValue() != null && parameter.getValue().length > 0) {
+                    DeviceCode = parameter.getValue()[0];
+                }
+            }
+        }
+
+        if (DeviceCode != null) {
+
+            if (!tokenReq.getGrantType().equals(Constants.DEVICE_FLOW_GRANT_TYPE)) {
+
+                throw new IdentityOAuth2Exception(DeviceErrorCodes.UNSUPPORTED_GRANT_TYPE);
+            } else {
+
+                DeviceFlowDO deviceFlowDO = DeviceFlowPersistenceFactory.getInstance().getDeviceFlowDAO()
+                        .getAuthenticationDetails(DeviceCode);
+                Date date = new Date();
+                deviceStatus = deviceFlowDO.getStatus();
+                if (Constants.NOT_EXIST.equals(deviceStatus)) {
+                    throw new IdentityOAuth2Exception(DeviceErrorCodes.INVALID_REQUEST);
+                } else if (Constants.EXPIRED.equals(deviceStatus) || isExpiredDeviceCode(deviceFlowDO, date)) {
```

Comment:
shouldn't we update the status of the device code if `isExpiredDeviceCode` is true?